### ISSUE TITLE
Vale.Spelling not reporting spelling errors when word contains capitalized letters after first letter

### ISSUE THREAD

**Issue by emteelb** [jdkato/txtlint#461]
OS = Linux
Install method = tar file download
Vale version 2.20.0

```
$ cat test.md
## My test Markdown file

iphone
iPhone
iPhonE
xxwrong
Xwrong
miistake
mistakE
miistakE
XXwrong
XXXwrong
XXXXwrong
ABCDwrong
xXwrong
xXXwrong
xxXwrong
xyzWrong
```

```
$ ./vale --minAlertLevel=warning test.md 

 test.md
 3:1  error  Did you really mean 'iphone'?   Vale.Spelling 
 6:1  error  Did you really mean 'xxwrong'?  Vale.Spelling 
 7:1  error  Did you really mean 'Xwrong'?   Vale.Spelling 
 8:1  error  Did you really mean             Vale.Spelling 
             'miistake'?                                   

✖ 4 errors, 0 warnings and 0 suggestions in 1 file.
```

```
$ ./vale --version
vale version 2.20.0
```

```
$ cat vale.ini
StylesPath = styles

MinAlertLevel = suggestion
#Vocab = Base

#Packages = proselint, write-good, alex, Readability

[*]
BasedOnStyles = Vale
```

**Comment by jdkato** [comment]
You can disable this by using your own custom filters, as stated in the [docs](https://vale.sh/docs/topics/styles/#spelling):

```yaml
extends: spelling
message: "Did you really mean '%s'?"
level: error
# This disables the built-in filters. If you omit this
# key or set it to false, custom filters (see below) are
# added on top of the built-in ones.
#
# By default, filters for acronyms, abbreviations, and
# numbers are included.
custom: true
# A "filter" is a regular expression specifying words
# to ignore during spell checking.
filters:
  # Ignore all words starting with 'py'.
  #
  # e.g., 'PyYAML'.
  - '[pP]y.*\b'
```

Although, IMO, these cases are usually better treated as style issues rather than spelling.

**Comment by emteelb** [comment]
Thanks for the reply and pointing me to the docs.

I am still not sure I understand why Vale.Spelling reports "miistake" as an error but not "miistakE". Is there an implicit filter in Vale.Spelling that ignores certain "word structures" like words that contain capital letters besides the first letter? The `defaultFilters` variable from line 16 from `spelling.go` perhaps?

And, if I understand your proposed solution, I should create a custom style rule that would flag as errors words that contain a capital letter anywhere in the word besides the first letter? Hopefully, if there are legitimate "camel case" words like "iPhone" or "OpenStack" that are either in the dictionary (or in `Vocab/MyProject/accept.txt`, or somewhere else?) these would not be flagged as errors by such a rule.

**Comment by jdkato** [comment]
> Is there an implicit filter in Vale.Spelling that ignores certain "word structures" like words that contain capital letters besides the first letter? 

Yes, because these words are:

- less likely to be true errors (case variations are an indication of intent more often then a mistake, especially in tech-related writing).
- very unlikely to be in the provided dictionary, even when valid (such as "OpenStack").

Combined, these factors lead to a lot of false positives.

> And, if I understand your proposed solution, I should create a custom style rule that would flag as errors words that contain a capital letter anywhere in the word besides the first letter?

No, I would suggest creating [`Vocab`(s)](https://vale.sh/docs/topics/vocab/) that list the expected form of accepted terms. For example, you could add "OpenStack" and then `Vale.Terms` will flag any instance that doesn't exactly match (something like "Openstack" becomes an error).

For these terms, I think this is more useful than spellcheck.

**Comment by emteelb** [comment]
Thanks. I am having some success with this method. However, I am having an issue trying to add a "phrase" to an `accept.txt` Vocab list.

If I am now straying too far from my original issue and it is worth starting a separate issue for discussion, happy to do that.

```
$ cat vale.ini
StylesPath = styles

MinAlertLevel = suggestion
Vocab = Test

#Packages = proselint, write-good, alex, Readability

[*]
BasedOnStyles = Vale
```
```
$ cat styles/Vocab/Test/accept.txt
TEST Suite
```
```
$ cat ts.md
TEST Suite is the correct spelling of the term
test suite
TEST suite
TEST-Suite
```
```
$ ./vale ts.md
✔ 0 errors, 0 warnings and 0 suggestions in 1 file.
```
I *can* get a match on the phrase if I put it into the `reject.txt` file.
```
$ cp styles/Vocab/Test/accept.txt styles/Vocab/Test/reject.txt
$ mv styles/Vocab/Test/accept.txt styles/Vocab/Test/accept.txt.hold
$ ./vale ts.md

 ts.md
 1:1  error  Avoid using 'TEST Suite'.  Vale.Avoid 

✖ 1 error, 0 warnings and 0 suggestions in 1 file.
```
So, I could put incorrect variations of my phrase into the reject list. (Or perhaps a substitution rule so that Vale reports the correct use as well as the error.)
```
$ cat styles/Vocab/Test/reject.txt
TEST-Suite
$ ./vale ts.md 

 ts.md
 4:1  error  Avoid using 'TEST-Suite'.  Vale.Avoid 

✖ 1 error, 0 warnings and 0 suggestions in 1 file.
```

However, a special problem happens if the first word in the phrase is in the accept list, on its own.
```
$ cat styles/Vocab/Test/accept.txt
TEST
$ ./vale ts.md 

 ts.md
 2:1  error  Use 'TEST' instead of 'test'.  Vale.Terms 

✖ 1 error, 0 warnings and 0 suggestions in 1 file.
```
Now Vale no longer reports the match from the reject list.

**Comment by jdkato** [comment]
> However, a special problem happens if the first word in the phrase is in the accept list, on its own.

You just need to be more specific about where the term can be used. If you just provide an entry of `TEST`, it can be used anywhere (including in `TEST-Suite`).

You need something like `TEST(?!-Suite)` in `accept.txt`.

**Comment by emteelb** [comment]
I appreciate your time working with me here. I tried that before my last comment but without success. Due to a typo in one of my accept file attempts, I discovered from a Vale error message that Vale seems to be adding case insensitivity, `(?i)`, to the accept "line" `TEST(?!\ Suite)`. So, I tried changing the accept line to `(?-i)TEST(?!\ Suite)` (switch off case insensitivity) but that still failed to cause Vale to report as errors the two lines that I would expect it to in my test file.

Accept file w/ an intentional typo to generate error message:
```
$ cat styles/Vocab/Test/accept.txt
TEST(?-\ Suite)
```
```
$ ./vale ts.md 
panic: error parsing regexp: unrecognized grouping construct: (?-\ in `(?i)TEST(?-\ Suite)`
[...]
```
With the typo corrected, and trying to switch off case insensitivity, Vale does not report the two occurrences of "test" in my test file as errors.
```
$ cat ts.md 
TEST Suite is the correct spelling of the term
test suite
TEST suite
TEST-Suite
test
```
```
$ cat styles/Vocab/Test/accept.txt 
(?-i)TEST(?!\ Suite)
```
```
$ ./vale ts.md 
✔ 0 errors, 0 warnings and 0 suggestions in 1 file.
```
I tried some variations of the line in my `accept.txt` Vocab file including: `TEST(?!\ Suite)`, `(?-i)TEST\b(?!\ Suite)`, `(?-i)TEST\ (?!Suite)`, just to experiment. None of those caused Vale to report occurrences of "test" as an error. My regex knowledge isn't phenomenal so I don't know if I just haven't hit on the correct regex, if this is an issue with Vale, or if it's an issue with my understanding of how the `accept.txt` Vocab file is supposed to work, or some combination of these.

**Comment by emteelb** [comment]
If it helps serve as a reminder, my end goal is for Vale to report as errors, variations of "TEST" and "TEST Suite" that aren't exactly that word or that phase. In my last comment, I removed the phrase from my accept file to try to just focus on hitting incorrect variations of "TEST" that would be a building block for adding another line in the accept file to generate errors for incorrect variations of "TEST Suite".

**Comment by jdkato** [comment]
> If it helps serve as a reminder, my end goal is for Vale to report as errors, variations of "TEST" and "TEST Suite" that aren't exactly that word or that phase.

This will do what you want:

```yaml
extends: substitution
message: Use '%s' instead of '%s'
ignorecase: true
level: error
swap:
  'test[ -]suite': TEST Suite
  'test(?! Suite)': TEST
```
![Screen Shot 2022-07-25 at 10 39 29 AM](https://user-images.githubusercontent.com/8785025/180840182-0472991b-07a3-4a21-ada2-e782bb1c42ec.png)

---

I'm having a hard time following the remainder of the issue. It seems like your efforts are *not* matching the stated goal: If you *don't* have case insensitivity (as used in the above rule), you'll never be able to find instances that differ by case (your stated goal).

**Comment by thomasten** [comment]
Recent versions of Vale do report some of the words from the initial post as spelling errors. Is this the desired behavior now? I find that especially annoying for plural forms of acronyms like `UUIDs`. If this is not a bug, how should I handle these? Add to accepted vocab?

**Comment by jdkato** [comment]
You have 3 options:

1. Maintain your own dictionary that you use in place of, or in addition to, the default one.
2. Maintain a vocab of accepted terms.
3. Use a [filter](https://vale.sh/docs/topics/styles/#ignoring-non-dictionary-words), such as the example in the docs:

```yml
extends: spelling
message: "Did you really mean '%s'?"
level: error
# This disables the built-in filters. If you omit this
# key or set it to false, custom filters (see below) are
# added on top of the built-in ones.
#
# By default, filters for acronyms, abbreviations, and
# numbers are included.
custom: true
# A "filter" is a regular expression specifying words
# to ignore during spell checking.
filters:
  # Ignore all words starting with 'py'.
  #
  # e.g., 'PyYAML'.
  - '[pP]y.*\b'
```
